File: content/examples/other/leetcode/problems/subtree-of-another-tree/example-2/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function serializePreorder(root) {
  const out = [];
  (function dfs(node) {
    if (!node) {
      out.push("#");
      return;
    }
    out.push(String(node.val));
    dfs(node.left);
    dfs(node.right);
  })(root);
  return out.join(",");
}

function isSubtreeSerialized(root, subRoot) {
  const a = serializePreorder(root);
  const b = serializePreorder(subRoot);
  return a.includes(b);
}

if (require.main === module) {
  const root = new TreeNode(3, new TreeNode(4, new TreeNode(1), new TreeNode(2)), new TreeNode(5));
  const sub = new TreeNode(4, new TreeNode(1), new TreeNode(2));
  console.log(isSubtreeSerialized(root, sub));
}

module.exports = { TreeNode, isSubtreeSerialized };

File: content/examples/other/leetcode/problems/subtree-of-another-tree/example-2/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def serialize_preorder(root):
    out = []

    def dfs(node):
        if node is None:
            out.append("#")
            return
        out.append(str(node.val))
        dfs(node.left)
        dfs(node.right)

    dfs(root)
    return ",".join(out)


def is_subtree_serialized(root, sub_root):
    return serialize_preorder(sub_root) in serialize_preorder(root)


if __name__ == "__main__":
    root = TreeNode(3, TreeNode(4, TreeNode(1), TreeNode(2)), TreeNode(5))
    sub = TreeNode(4, TreeNode(1), TreeNode(2))
    print(is_subtree_serialized(root, sub))
